pydantic/analyzed_type.py

```diff
@@ -0,0 +1,241 @@
+
+import sys
+from typing import TYPE_CHECKING, Any, Dict, Generic, Iterable, Set, TypeVar, Union, overload
+
+from pydantic_core import CoreConfig, CoreSchema, SchemaSerializer, SchemaValidator, core_schema
+from typing_extensions import Literal
+
+from pydantic.config import ConfigDict
+from pydantic.json_schema import DEFAULT_REF_TEMPLATE, GenerateJsonSchema
+
+from ._internal import (
+    _generate_schema,
+    _typing_extra,
+)
+
+T = TypeVar('T')
+
+if TYPE_CHECKING:
+    # should be `set[int] | set[str] | dict[int, IncEx] | dict[str, IncEx] | None`, but mypy can't cope
+    IncEx = Union[Set[int], Set[str], Dict[int, Any], Dict[str, Any]]
+
+
+def _get_schema(type_: Any, config: CoreConfig | None, parent_depth: int) -> CoreSchema:
+    arbitrary_types = bool((config or {}).get('arbitrary_types_allowed', False))
+    local_ns = _typing_extra.parent_frame_namespace(parent_depth=parent_depth)
+    # BaseModel uses it's own __module__ to find out where it was defined
+    # and then look for symbols to resolve forward references in those globals
+    # On the other hand this function can be called with arbitrary objects,
+    # including type aliases where __module__ (always `typing.py`) is not useful
+    # So instead we look at the globals in our parent stack frame
+    # This works for the case where this function is called in a module that
+    # has the target of forward references in its scope but
+    # does not work for more complex cases
+    # for example, take the following:
+    #
+    # a.py
+    # ```python
+    # from typing import List, Dict
+    # IntList = List[int]
+    # OuterDict = Dict[str, 'IntList']
+    # ```
+    #
+    # b.py
+    # ```python
+    # from pydantic import AnalyzedType
+    # from a import OuterDict
+    # IntList = int  # replaces the symbol the forward reference is looking for
+    # v = AnalyzedType(OuterDict)
+    # v({"x": 1})  # should fail but doesn't
+    # ```
+    #
+    # If OuterDict were a BaseModel this would work because it would resolve
+    # the forward reference within the `a.py` namespace.
+    # But `AnalyzedType(OuterDict)`
+    # can't know what module OuterDict came from.
+    # In other words, the assumption that _all_ forward references exist in the
+    # module we are being called from is not technically always true
+    # Although most of the time it is and it works fine for recursive models and such/
+    # BaseModel's behavior isn't perfect either and _can_ break in similar ways,
+    # so there is no right or wrong between the two.
+    # But at the very least this behavior is _subtly_ different from BaseModel's.
+    global_ns = sys._getframe(max(parent_depth - 1, 1)).f_globals.copy()
+    global_ns.update(local_ns or {})
+    gen = _generate_schema.GenerateSchema(arbitrary_types=arbitrary_types, types_namespace=global_ns, typevars_map={})
+    return gen.generate_schema(type_)
+
+
+def _translate_config(config: ConfigDict) -> core_schema.CoreConfig:
```

Comment:
Is this the same as `_internal/_generate_schema.py::generate_config`?

Reply:
if it's not the same now because `config` really is partial, we should add a TODO to replace it when we update config logic.